crates/matrix-sdk/src/event_cache/mod.rs

```diff
@@ -501,17 +528,60 @@ impl RoomEventCache {
         None
     }
 
+    /// Try to find an event by id in this room, along with all relations.
+    pub async fn event_with_relations(
+        &self,
+        event_id: &EventId,
+    ) -> Option<(SyncTimelineEvent, Vec<SyncTimelineEvent>)> {
+        let mut relation_events = Vec::new();
+
+        let cache = self.inner.all_events_cache.read().await;
+        if let Some((_, event)) = cache.events.get(event_id) {
+            Self::related_event_ids_for_event(&cache, event_id, &mut relation_events);
+            Some((event.clone(), relation_events))
+        } else {
+            None
+        }
+    }
+
+    /// Looks for related event ids for the passed event id, and appends them to
+    /// the `results` parameter. Then it'll recursively get the related
+    /// event ids for those too.
+    fn related_event_ids_for_event(
+        cache: &RwLockReadGuard<'_, AllEventsCache>,
+        event_id: &EventId,
+        results: &mut Vec<SyncTimelineEvent>,
```

Comment:
Do you think it would be ok to have `results` be a `HashSet` or `BTreeSet` instead? (and return that from the `event_with_relations` function)

It would avoid the linear search below, and pretty sure consumers don't care about the order; after all we're returning an arbitrary order of related events here.

Reply:
A couple of questions about that: 

- Wouldn't that mean we'd have to implement either `Hash` or `Ord` for `SyncTimelineEvent`? And if so, what would the hash value or the ordering be based on?
- Are we sure ordering doesn't matter for the `TimelineEventHandler`? `rfind_event_by_id` is used in some places to look for related events, and that will only work if the event was already added to `self.items`, right?

Reply:
- You're right about `Hash` for both data structures, if we only use `HashSet` we don't need `Ord` (we'd need it for `BTreeSet` though). If it's not implemented already, let's forget about that idea.
- The "only" thing that matters with respect to ordering is that the related-to event is added before the relation events, if I understand correctly; otherwise you're right it would be problematic.

Reply:
We already discussed this in person, but just so we track the conclusions:
- We don't have `Hash` implemented and while we could generate it from the contents, it's probably not worth it.
- Using a `Set` could be a problem because with this code we'd also add transitively related events to it (a related event for a related event) and that may break the current code for handling timeline events since the original event for a related event must be already added to the timeline and we need the topological order to ensure that.